### PR TITLE
Gramine ratls wrapper

### PR DESCRIPTION
**Pull request by lonerapier**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -54,89 +56,164 @@ package main
...
+		C.set_g_verify_isv_svn(false)
+	} else {
+		C.set_g_verify_isv_svn(true)
+		C.set_g_expected_isv_svn(unsafe.Pointer(&isv_svn[0]), C.size_t(len(isv_svn)))
 	}
+}
+
+func RATLSVerifyDer(certDER, mrenclave, mrsigner, isv_prod_id, isv_svn []byte) error {
+	// set verify callback
+	C.ra_tls_set_measurement_callback_wrapper(ra_tls_set_measurement_callback_f)
```
Comment (fnerdman) [inline main.go]: Shouldn't this be part of initialization?

File: main.go [hunk 2]
```diff
@@ -54,89 +56,164 @@ package main
...
 	defer C.free(unsafe.Pointer(certDER_sym))
-	ret := C.ra_tls_verify_callback_der_wrapper(ra_tls_verify_callback_der_f, (*C.uchar)(certDER_sym), cert_size)
 
-	// convert der cert to C objects
-	// copy attestation consumables to static c objects
-	// call measurement callback
+	ret := C.ra_tls_verify_callback_der_wrapper(ra_tls_verify_callback_der_f, (*C.uchar)(certDER_sym), cert_size)
 
 	fmt.Println("Success: ", ret)
+	if ret != 0 {
```
Comment (fnerdman) [inline main.go]: Return more verbose error message, for return values you can check: https://github.com/gramineproject/gramine/blob/master/tools/sgx/ra-tls/ra_tls_verify_dcap.c#L98

Unfortunately [the verification result is abstracted away](https://github.com/gramineproject/gramine/blob/master/tools/sgx/ra-tls/ra_tls_verify_dcap.c#L165-L185). We might want to work around this at some point.

File: main.go [hunk 3]
```diff
@@ -54,89 +56,164 @@ package main
...
 	certDER_sym := C.CBytes(certDER)
 	defer C.free(unsafe.Pointer(certDER_sym))
-	ret := C.ra_tls_verify_callback_der_wrapper(ra_tls_verify_callback_der_f, (*C.uchar)(certDER_sym), cert_size)
 
-	// convert der cert to C objects
-	// copy attestation consumables to static c objects
-	// call measurement callback
+	ret := C.ra_tls_verify_callback_der_wrapper(ra_tls_verify_callback_der_f, (*C.uchar)(certDER_sym), cert_size)
 
 	fmt.Println("Success: ", ret)
```
Comment (fnerdman) [inline main.go]: Use log package and log levels to output debug messages

File: main.go [hunk 4]
```diff
@@ -54,89 +56,164 @@ package main
...
+	}
+
+	block, _ := pem.Decode(cert)
+	if block == nil {
+		return errors.New("failed to decode PEM data")
+	}
+
+	certificate, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return errors.New("failed to create cert")
```
Comment (fnerdman) [inline main.go]: You could just pass on the error here.

File: go.mod [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module ratls
```
Comment (fnerdman) [inline go.mod]: I would opt for `gramine-ratls`,  the package name and the file name should also be named `gramine-ratls`

File: main.go [hunk 6]
```diff
@@ -54,89 +56,164 @@ package main
...
+	}
+
+	return RATLSVerifyDer(certificate.Raw, mrenclave, mrsigner, isv_prod_id, isv_svn)
+}
+
+func main() {
```
Comment (fnerdman) [inline main.go]: The repo should only be library and not provide a main() function

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+var certFilePath = "tls/tlscert.der"
+
+func Test_RATLSVerifyDer(t *testing.T) {
+	certFile, err := os.ReadFile(certFilePath)
+	if err != nil {
+		t.Error("error in opening file")
+	}
+
+	err = RATLSVerifyDer(certFile, nil, nil, nil, nil)
```
Comment (fnerdman) [inline main_test.go]: Add more test cases,

i.e. testing measurements,

Using [gramine documented](https://gramine.readthedocs.io/en/stable/attestation.html) env variables

**Review by fnerdman — CHANGES_REQUESTED** [review]
Apart from the comments in the code I would move the test tls cert to a `test` folder

**Comment by lonerapier** [comment]
> Apart from the comments in the code I would move the test tls cert to a test folder

Go's [standard](https://stackoverflow.com/questions/19200235/golang-tests-in-sub-directory) is to have the test file along with the source code with `_test` suffix.

What is the reason behind putting it in a sub-directory?

File: go.mod (hunk 5 above)
Comment (lonerapier) [reply go.mod]: Golang doesn't allow `-` in the package name of the source file, will rename it to `gramine_ratls`. Does this work?

**Comment by fnerdman** [comment]
It is only the tls certificate folder that should be moved to a `test` folder. It should be clear that the files in the tls folder are test consumables and not part of the code.

**Comment by fnerdman** [comment]
* Please add a Readme that documents the methods
* please use golangs official logging library

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+main
```
Comment (fnerdman) [inline .gitignore]: No more need for this as we're not getting an executable anymore

File: gramine_ratls_test.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+var certDer []byte
+var certPem []byte
+
+// RA-TLS measurements args
+var mrenclave []byte
+var mrsigner []byte
+var isv_prod_id []byte
+var isv_svn []byte
+
+func setup() {
```
Comment (fnerdman) [inline gramine_ratls_test.go:28]: We should probably set the env var `RA_TLS_ALLOW_OUTDATED_TCB_INSECURE` here.

**Review by fnerdman — CHANGES_REQUESTED** [review]

File: gramine_ratls_test.go (hunk 9 above)
Comment (lonerapier) [reply gramine_ratls_test.go:28]: Isn't that supposed to be insecure practice?

File: README.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
+# Gramine-RATLS-Wrapper
+
+Gramine-RATLS-Wrapper is a Go wrapper on top of the [Gramine](https://github.com/gramineproject/gramine) [Remote Attestation TLS](https://github.com/gramineproject/gramine/tree/master/tools/sgx/ra-tls) (RATLS) library. Forwards the certificate and SGX quote to the DCAP verification service. It also verifies SGX specific measurement variables that are used to verify the identity of the enclave, namely:
+
+1. `MRENCLAVE`: hex string that verifes attesting enclave.
+2. `MRSIGNER`: hex string that verifies the key used to sign the enclave.
+3. `ISV_PROD_ID`: decimal string that verifes the valid product ID.
+4. `ISV_SVN`: decimal string metadata variable that specifes valid version number.
+
+More info about the core library can be found at the official gramine [docs](https://github.com/gramineproject/gramine).
```
Comment (fnerdman) [inline README.md]: Is this the wrong link? -> https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface

File: README.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+		fmt.Printf("Certificate verification failed: %v\n", err)
+	} else {
+		fmt.Println("Certificate verified successfully.")
+	}
+}
+```
+
+In this example, we assume that we have a RATLS certificate in a byte slice called `cert`, and that we have the `mrenclave`, `mrsigner`, `isv_prod_id`, and `isv_svn` values in byte slices. We call the `RATLSVerifyDer` function to verify the certificate. If verification fails, we print an error message. If verification succeeds, we print a success message.
+
+Similarly, we can call the `RATLSVerify` function to verify a PEM-encoded RATLS certificate. Note that in both cases, the `cert` parameter must be a DER-encoded or PEM-encoded RATLS certificate, depending on which function is being called.
```
Comment (fnerdman) [inline README.md]: Could be left out, was already mentioned in the PEM method description.

File: README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+The wrapper provides two functions for verifying RATLS certificates:
+
+### RATLSVerifyDer
+
+```go
+func RATLSVerifyDer(cert, mrenclave, mrsigner, isv_prod_id, isv_svn []byte) error
+```
+
+RATLSVerifyDer verifies a DER-encoded RATLS certificate using the given `mrenclave`, `mrsigner`, `isv_prod_id`, and `isv_svn`. If the certificate is valid, this function returns `nil`. If the certificate is invalid, this function returns an error describing the validation failure.
```
Comment (fnerdman) [inline README.md]: Please mention that the value of the measurements can be set to nil, which means their values will be ignore.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+		fmt.Println("Certificate verified successfully.")
+	}
+}
+```
+
+In this example, we assume that we have a RATLS certificate in a byte slice called `cert`, and that we have the `mrenclave`, `mrsigner`, `isv_prod_id`, and `isv_svn` values in byte slices. We call the `RATLSVerifyDer` function to verify the certificate. If verification fails, we print an error message. If verification succeeds, we print a success message.
+
+Similarly, we can call the `RATLSVerify` function to verify a PEM-encoded RATLS certificate. Note that in both cases, the `cert` parameter must be a DER-encoded or PEM-encoded RATLS certificate, depending on which function is being called.
+
+> Note that enclave measurement arguments can be optional depending on the measurement callback defined as per usage.
```
Comment (fnerdman) [inline README.md]: I would put this into the method description, as already mentiond above

File: README.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	err := gramine_ratls.RATLSVerifyDer(cert, mrenclave, mrsigner, isv_prod_id, isv_svn)
+	if err != nil {
+		fmt.Printf("Certificate verification failed: %v\n", err)
+	} else {
+		fmt.Println("Certificate verified successfully.")
+	}
+}
+```
+
+In this example, we assume that we have a RATLS certificate in a byte slice called `cert`, and that we have the `mrenclave`, `mrsigner`, `isv_prod_id`, and `isv_svn` values in byte slices. We call the `RATLSVerifyDer` function to verify the certificate. If verification fails, we print an error message. If verification succeeds, we print a success message.
```
Comment (fnerdman) [inline README.md]: Describing the If-statement is going to much into detail

File: README.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+## Usage
+
+To use RA-TLS wrapper, import the `gramine_ratls` package and call the RATLSVerifyDer or RATLSVerify functions depending on the certificate:
+
+```go
+func main() {
+	// Assume that we have a RATLS certificate in a byte slice called "cert"
+	// Assume that we have the mrenclave, mrsigner, isv_prod_id, and isv_svn values in byte slices
+	err := gramine_ratls.RATLSVerifyDer(cert, mrenclave, mrsigner, isv_prod_id, isv_svn)
```
Comment (fnerdman) [inline README.md]: maybe set measurements with the test values. This gives us an idea of how they need to look.

**Review by fnerdman — CHANGES_REQUESTED** [review]

File: gramine_ratls.go [hunk 16]
```diff
@@ -106,19 +106,21 @@ import (
 	"encoding/pem"
 	"errors"
 	"fmt"
+	"os"
 	"unsafe"
 
-	"github.com/sirupsen/logrus"
```
Comment (fnerdman) [inline gramine_ratls.go]: Can you revert to logrus? You were right - with the official log package not supporting log levels, it doesn't make much sense to use it. logrus is beeing used by Docker so I'm confident it is being monitored closely for security issues.

File: README.md (hunk 10 above)
Comment (lonerapier) [reply README.md]: oops. thanks, changed.

**Review by fnerdman — APPROVED** [review]

**Pull request merged**